src/dbnode/persist/fs/seek_manager.go

```diff
@@ -286,15 +343,188 @@ func (m *seekerManager) Return(shard uint32, start time.Time, seeker ConcurrentD
+// released (in this case the Seeker / files for the previous volume) and the resources associated
+// with the new lease should have been acquired (the seeker for the provided volume).
+//
+// Practically speaking, the goal of this function is to open a new seeker for the latest volume and
+// then "hot-swap" it so that by the time this function returns there are no more outstanding reads
+// using the old seekers, all the old seekers have been closed, and all subsequent reads will use the
+// seekers associated with the latest volume.
+//
+// The bulk of the complexity of this function is caused by the desire to avoid the hot-swap from
+// causing any latency spikes. To accomplish this, the following is performed:
+//
+//   1. Open the new seeker outside the context of any locks.
+//   2. Acquire a lock on the seekers that need to be swapped and rotate the existing "active" seekers
+//      to be "inactive" and set the newly opened seekers as "active". This operation is extremely cheap
+//      and ensures that all subsequent reads will use the seekers for the latest volume instead of the
+//      previous. In addition, this phase also creates a waitgroup for the inactive seekers that will be
+//      be used to "wait" for all of the existing seekers that are currently borrowed to be returned.
+//   3. Release the lock so that reads can continue uninterrupted and call waitgroup.Wait() to wait for all
+//      the currently borrowed "inactive" seekers (if any) to be returned.
+//   4. Every call to Return() for an "inactive" seeker will check if it's the last borrowed inactive seeker,
+//      and if so, will close all the inactive seekers and call wg.Done() which will notify the goroutine
+//      running the UpdateOpenlease() function that all inactive seekers have been returned and closed at
+//      which point the function will return sucessfully.
 func (m *seekerManager) UpdateOpenLease(
 	descriptor block.LeaseDescriptor,
 	state block.LeaseState,
 ) (block.UpdateOpenLeaseResult, error) {
-	// TODO(rartoul): This is a no-op for now until the logic for swapping out seekers is written.
-	// Also make sure that this function is a proper no-op if the SeekerManager state has been
-	// been switched to closed.
-	return block.NoOpenLease, nil
+	m.Lock()
+	if m.status != seekerManagerOpen {
+		m.Unlock()
+		return 0, errUpdateOpenLeaseSeekerManagerNotOpen
+	}
+
+	if m.isUpdatingLease {
+		// This guard is a little overly aggressive. In practice, the algorithm remains correct even in the presence
+		// of concurrent UpdateOpenLease() calls as long as they are for different shard/blockStart combinations.
+		// However, the calling code currently has no need to call this method concurrently at all so use the
+		// simpler check for now.
+		m.Unlock()
+		return 0, errConcurrentUpdateOpenLeaseNotAllowed
+	}
+
+	if !m.namespace.Equal(descriptor.Namespace) {
+		m.Unlock()
+		return block.NoOpenLease, nil
+	}
+	m.isUpdatingLease = true
+	m.Unlock()
```

Comment:
nit: Could the code above be grouped into a `startUpdateOpenLease() error`?

That would be able to use `m.Lock()` and `defer`:
```go
func (m *seekerManager) UpdateOpenLease(
	descriptor block.LeaseDescriptor,		descriptor block.LeaseDescriptor,
	state block.LeaseState,		state block.LeaseState,
) (block.UpdateOpenLeaseResult, error) {	) (block.UpdateOpenLeaseResult, error) {
	m.Lock()
	if m.status != seekerManagerOpen {
		m.Unlock()
		return 0, errUpdateOpenLeaseSeekerManagerNotOpen
	}

 	if m.isUpdatingLease {
		// This guard is a little overly aggressive. In practice, the algorithm remains correct even in the presence
		// of concurrent UpdateOpenLease() calls as long as they are for different shard/blockStart combinations.
		// However, the calling code currently has no need to call this method concurrently at all so use the
		// simpler check for now.
		m.Unlock()
		return 0, errConcurrentUpdateOpenLeaseNotAllowed
	}

 	if !m.namespace.Equal(descriptor.Namespace) {
		m.Unlock()
		return block.NoOpenLease, nil
	}
	m.isUpdatingLease = true
	m.Unlock()

	// .... other code
```

Becomes:
```go
func (m *seekerManager) startUpdateOpenLease(
	descriptor block.LeaseDescriptor,
) error {
	m.Lock()
	defer m.Unlock()
	if m.status != seekerManagerOpen {
		return 0, errUpdateOpenLeaseSeekerManagerNotOpen
	}

 	if m.isUpdatingLease {
		// This guard is a little overly aggressive. In practice, the algorithm remains correct even in the presence
		// of concurrent UpdateOpenLease() calls as long as they are for different shard/blockStart combinations.
		// However, the calling code currently has no need to call this method concurrently at all so use the
		// simpler check for now.
		return 0, errConcurrentUpdateOpenLeaseNotAllowed
	}

 	if !m.namespace.Equal(descriptor.Namespace) {
		return block.NoOpenLease, nil
	}
	m.isUpdatingLease = true
	return nil
}

func (m *seekerManager) UpdateOpenLease(
	descriptor block.LeaseDescriptor,
	state block.LeaseState,
) (block.UpdateOpenLeaseResult, error) {
	if err := m.startUpdateOpenLease(); err != nil {
		return err
	}
	defer func() {
		m.Lock()
		m.isUpdatingLease = false
		m.Unlock()
	}()

	// .... other code
```

Reply:
yeah sure this is better